packages/agent/src/agent/http/index.ts

```diff
@@ -418,13 +425,98 @@ export class HttpAgent implements Agent {
     };
   }
 
+  async #requestAndRetryQuery(
+    args: {
+      canister: string;
+      transformedRequest: HttpAgentRequest;
+      body: ArrayBuffer;
+      requestId: RequestId;
+    },
+    tries = 0,
+  ): Promise<ApiQueryResponse> {
+    const { canister, transformedRequest, body, requestId } = args;
+    let response: ApiQueryResponse;
+    // Make the request and retry if it throws an error
+    try {
+      const fetchResponse = await this._fetch(
+        '' + new URL(`/api/v2/canister/${canister}/query`, this._host),
+        {
+          ...this._fetchOptions,
+          ...transformedRequest.request,
+          body,
+        },
+      );
+      const queryResponse: QueryResponse = cbor.decode(await fetchResponse.arrayBuffer());
+      response = {
+        ...queryResponse,
+        httpDetails: {
+          ok: fetchResponse.ok,
+          status: fetchResponse.status,
+          statusText: fetchResponse.statusText,
+          headers: httpHeadersTransform(fetchResponse.headers),
+        },
+        requestId,
+      };
+    } catch (error) {
+      if (tries < this._retryTimes) {
+        this.log.warn(
+          `Caught exception while attempting to make query:\n` +
+            `  ${error}\n` +
+            `  Retrying query.`,
+        );
+        return await this.#requestAndRetryQuery(args, tries + 1);
+      }
+      throw error;
+    }
+
+    const timestamp = response.signatures?.[0]?.timestamp;
+
+    // Skip watermark verification if the user has set verifyQuerySignatures to false
+    if (!this.#verifyQuerySignatures) {
+      return response;
+    }
+
+    if (!timestamp) {
+      throw new Error(
+        'Timestamp not found in query response. This suggests a malformed or malicious response.',
+      );
+    }
+
+    // Convert the timestamp to milliseconds
+    const timeStampInMs = Number(BigInt(timestamp) / BigInt(1_000_000));
+
+    this.log('watermark and timestamp', {
+      waterMark: this.waterMark,
+      timestamp: timeStampInMs,
+    });
+
+    // If the timestamp is less than the watermark, retry the request up to the retry limit
+    if (timestamp && Number(this.waterMark) > timeStampInMs) {
```

Comment:
nit: do we even need to evaluate `timestamp` at that point? We checked it in line 479, right?

Reply:
Yes - the error above makes this redundant. I'll simplify